universal.c

```diff
@@ -593,11 +593,21 @@ XS(XS_utf8_upgrade)
         croak_xs_usage(cv, "sv");
     else {
         SV * const sv = ST(0);
-        STRLEN	RETVAL;
+        STRLEN	RETVAL = 0;
         dXSTARG;
 
-        RETVAL = sv_utf8_upgrade(sv);
-        XSprePUSH; PUSHi((IV)RETVAL);
+        XSprePUSH;
+        if (LIKELY(sv)) {
+            SvGETMAGIC(sv);
+            if (LIKELY(SvOK(sv))) {
+                RETVAL = sv_utf8_upgrade_nomg(sv);
+                PUSHi( (IV) RETVAL);
+            } else {
+                PUSHs(newSV(0));
```

Comment:
This leaks, since the SV isn't made mortal (tested with `./perl -e 'my $x; while (1) { utf8::upgrade($x); }'`)

That isn't a problem for PUSHi() since it uses TARG.

You can simply use `XSRETURN_UNDEF;` here, and immediately below.